presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java

```diff
@@ -57,6 +64,34 @@ private ResourceUtil()
 
     public static Session createSessionForRequest(HttpServletRequest servletRequest)
     {
+        String realUser = null;
+        Principal authenticatedPrincipal = servletRequest.getUserPrincipal();
+        if (authenticatedPrincipal != null) {
+            String delegationJsonValue = servletRequest.getHeader(PRESTO_DELEGATION);
+            if (delegationJsonValue != null) {
+                //TODO check if the authenticatedPrincipal has delegation privilege
+                try {
+                    JsonCodec<Delegation> jsonCodec = jsonCodec(Delegation.class);
+                    Delegation delegation = jsonCodec.fromJson(delegationJsonValue);
+                    DelegationIdentity identity = delegation.getIdentity();
+                    if (identity != null) {
+                        if (identity.getType() == UNIX_PRINCIPAL) {
+                            realUser = identity.getName();
+                        }
+                        else if (identity.getType() == KERBEROS_PRINCIPAL) {
+                            realUser = (new KerberosPrincipal(identity.getName())).getName();
```

Comment:
The delegation principal will be used for authorization. It may make sense to add a new member 'delegatedPrincipal' to Session.